content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+#3  0x0000000000433b4a in runtime.stoplockedm () at /usr/local/go/src/runtime/proc.go:2165
+#4  0x0000000000435279 in runtime.schedule () at /usr/local/go/src/runtime/proc.go:2565
+#5  0x00000000004353fe in runtime.park_m (gp=0xc000066d80) at /usr/local/go/src/runtime/proc.go:2676
+#6  0x000000000045ae1b in runtime.mcall () at /usr/local/go/src/runtime/asm_amd64.s:299
+#7  0x000000000045ad39 in runtime.rt0_go () at /usr/local/go/src/runtime/asm_amd64.s:201
+#8  0x0000000000000000 in ?? ()
+```
+
+Honestly, I’m not a GDB expert, but it seems that Go runtime is putting threads to sleep. *But why?*
+
+Debugging a live process is “fun”, also let’s grab a coredump of the process to analyse it offline. We can do it with GDB's `gcore` command. The core file will be saved as `core.<process_id>` in the current working directory. Note, even for our simple server, the file will be pretty big. It’s likely to be huge for production service.
+
+```
+(gdb) gcore
+Saved corefile core.1628
+(gdb) quit
+
+= du -h core.1628
+1.2G    core.1628
+```
+
+With coredump saved, we're no longer required to keep the process running. Feel free to “kill -9” it.
+
+*For more info on the topic of debugging with GDB, check out Go’s own "[Debugging Go Code with GDB][2]".*
+
+## Enter Delve, Debugger for Go
+
+[Delve][] is the debugger for Go programs. It is similar to GDB but is aware of Go's runtime, data structures and other internals of the language.
+
+I highly recommend a talk by Alessandro Arzilli “[Internal Architecture of Delve, a Debugger For Go](https://www.youtube.com/watch?v=IKnTr7Zms1k)” from GopherCon EU 2018 if you're interested to know about Delve's internals.
+
+Delve is written in Go so installing it as simple as running:
+
+```
+= go get -u github.com/derekparker/delve/cmd/dlv
+```
+
+With Delve installed we can start analysing the core file by running `dlv core <path to service binary> <core file>`. We start by listing all goroutines that were running when the coredump was taken. Delve's `goroutines` command does exactly this:
+
+```
+= dlv core example/server/server core.1628
+
+(dlv) goroutines
+  ···
+  Goroutine 4611 - User: /vagrant/example/server/metrics.go:113 main.(*Metrics).CountS (0x703948)
+  Goroutine 4612 - User: /vagrant/example/server/metrics.go:113 main.(*Metrics).CountS (0x703948)
+  Goroutine 4613 - User: /vagrant/example/server/metrics.go:113 main.(*Metrics).CountS (0x703948)
+
+```
+
+*Unfortunately, in a real case scenario, the list can be so big, it doesn't even fit into terminal's scroll buffer. Remember that the server spawns a goroutine for each incoming request, so “goroutines” command has shown us a list of almost a million items. Let's pretend that we faced exactly this, and think of a way to work through this situation.*
+
+We can run Delve in the "headless" mode to interact with the debugger via it's [JSON-RPC API](https://github.com/derekparker/delve/tree/master/Documentation/api).
```

Comment:
Typo here, "it's" is the short for of "it is", the possessive form of "it" is "its", https://www.lawlessenglish.com/english-mistakes/its/
```suggestion
We can run Delve in the "headless" mode to interact with the debugger via its [JSON-RPC API](https://github.com/derekparker/delve/tree/master/Documentation/api).
```